pkg/client/query.go

```diff
@@ -0,0 +1,244 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	maxComparisonDelta = 0.001
+
+	comparisonSuccess = "success"
+	comparisonFailed  = "fail"
+
+	querySkipped = "skipped"
+	querySuccess = "success"
+	queryFailed  = "fail"
+)
+
+type QueryClientConfig struct {
+	URL string
+
+	// The tenant ID to use to push metrics to Cortex.
+	UserID string
+
+	QueryInterval time.Duration
+	QueryTimeout  time.Duration
+	QueryMaxAge   time.Duration
+
+	ExpectedSeries        int
+	ExpectedWriteInterval time.Duration
+}
+
+type QueryClient struct {
+	cfg       QueryClientConfig
+	client    v1.API
+	startTime time.Time
+	logger    log.Logger
+
+	// Metrics.
+	queriesTotal         *prometheus.CounterVec
+	resultsComparedTotal *prometheus.CounterVec
+}
+
+func NewQueryClient(cfg QueryClientConfig, logger log.Logger, reg prometheus.Registerer) *QueryClient {
+	var rt http.RoundTripper = &http.Transport{}
```

Comment:
Should we use some sane defaults instead of empty `http.Transport`? (Eg. similar to those in `http.DefaultTransport`). Empty Transport has some weird defaults (like no timeouts, or no proxy support)